rust/src/bs_safe.rs

```diff
@@ -0,0 +1,599 @@
+        &self,
+        theta_unc: &[f64],
+        include_tp: bool,
+        include_gq: bool,
+        out: &mut [f64],
+        rng: Option<&mut Rng<R>>,
+    ) -> Result<()> {
+        let n = self.param_unc_num();
+        assert_eq!(
+            theta_unc.len(),
+            n,
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        let out_n = self.param_num(include_tp, include_gq);
+        assert_eq!(
+            out.len(),
+            out_n,
+            "Argument 'out' must be the same size as the number of parameters!"
+        );
+
+        if include_gq {
+            assert!(
+                rng.is_some(),
+                "Rng was not provided even though generated quantities are requested."
+            );
+        }
+
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.lib.borrow().0.bs_param_constrain(
+                self.model.as_ptr(),
+                include_tp as c_int,
+                include_gq as c_int,
+                theta_unc.as_ptr(),
+                out.as_mut_ptr(),
+                rng.map(|rng| rng.rng.as_ptr()).unwrap_or(null_mut()),
+                err.as_ptr(),
+            )
+        };
+
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Set the sequence of unconstrained parameters based on the
+    /// specified constrained parameters, and return a return code of 0
+    /// for success and -1 for failure.  Parameter order is as declared
+    /// in the Stan program, with multivariate parameters given in
+    /// last-index-major order.
+    ///
+    /// # Arguments
+    /// `theta`: The vector of constrained parameters
+    ///
+    /// `theta_unc` Vector of unconstrained parameters
+    pub fn param_unconstrain(&self, theta: &mut [f64], theta_unc: &mut [f64]) -> Result<()> {
+        assert_eq!(
+            theta_unc.len(),
+            self.param_unc_num(),
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        assert_eq!(
+            theta.len(),
+            self.param_num(false, false),
+            "Argument 'out' must be the same size as the number of parameters!"
+        );
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.lib.borrow().0.bs_param_unconstrain(
+                self.model.as_ptr(),
+                theta.as_ptr(),
+                theta_unc.as_mut_ptr(),
+                err.as_ptr(),
+            )
+        };
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Map a constrained point in json format to the unconstrained space.
+    ///
+    /// The parameter order in the JSON is as declared in the Stan program, with
```

Comment:
This is not quite correct. The JSON doesn't need to do things like flatten into last-index-major order. It's essentially the same kind of thing you'd use for initialization in Stan or for the data argument, just for the parameters block.

Reply:
I mostly copied that from `bridgestan.h`:

```
/**
 * Set the sequence of unconstrained parameters based on the JSON
 * specification of the constrained parameters, and return a return
 * code of 0 for success and -1 for failure.  Parameter order is as
 * declared in the Stan program, with multivariate parameters given
 * in last-index-major order. The JSON schema assumed is fully
 * defined in the *CmdStan Reference Manual*.
 *
 * @param[in] m pointer to model structure
 * @param[in] json JSON-encoded constrained parameters
 * @param[out] theta_unc sequence of unconstrained parameters
 * @param[out] error_msg a pointer to a string that will be allocated if there
 * is an error. This must later be freed by calling `bs_free_error_msg`.
 * @return code 0 if successful and code -1 if there is an exception
 * in the underlying Stan code
 */
```
Do you have a better description in mind? I can fix both at the same time then.

Reply:
Oh that's definitely a copy/paste error on our part. 
The sentence `Parameter order is as declared in the Stan program, with multivariate parameters given in last-index-major order.` should just be omitted. If anything needs to be said, something like "A value for each parameter in the Stan program should be provided, with dimensions and size corresponding to the Stan program declarations"

Reply:
I changed the docstring in rust and C (the other languages didn't have that sentence)